src/libusb.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2017 Assured Information Security, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+/**
+ * @file   libusb.c
+ * @author Troy Crosley <crosleyt@ainfosec.com>
+ * @date   Fri Feb 03 10:53:45 2017
+ *
+ * @brief  libusb interaction
+ *
+ * Functions that usb libusb to find more about the device type
+ */
+
+#include "project.h"
+
+#define BMATTRIBUTES_BULK               0x02
+#define BENDPOINTADDRESS_IN             0x80
+#define TYPICAL_NIC_PACKET_SIZE         0x0200
+#define MAX_ENDPOINTS                   1000
+
+static bool
+libusb_is_ethernet_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == COMMUNICATIONS_CLASS &&
+    interface->bInterfaceSubClass == ETHERNET_NETWORKING_SUBCLASS);
+}
+
+static bool
+libusb_is_wireless_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == WIRELESS_CLASS &&
+    (interface->bInterfaceSubClass != RADIO_FREQUENCY_SUBCLASS ||
+    interface->bInterfaceProtocol != BLUETOOTH_PROTOCOL));
+}
+
+static bool
+libusb_is_bluetooth_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == WIRELESS_CLASS &&
+    interface->bInterfaceSubClass == RADIO_FREQUENCY_SUBCLASS &&
+    interface->bInterfaceProtocol == BLUETOOTH_PROTOCOL);
+}
+
+/**
+ * Identify vendor-specific interfaces with endpoints that support
+ * bulk data transfers both in and out with a packet size of at least
+ * TYPICAL_NIC_PACKET_SIZE (512). Though all the USB NICs that were
+ * tested had such an endpoint and none of the other USB devices did,
+ * It is possible that checking for NICs in this manner will result
+ * in false positives or negatives.
+ */
+static bool
+libusb_is_unknown_bulk_transfer_interface(struct usb_interface_descriptor *interface)
+{
```

Comment:
It would if you allowed NICs but not keyboards and the rule for allowing NICs had a lower number than the rule for denying keyboards (which it shouldn't).